src/ddspdrum/module.py

```diff
@@ -305,20 +326,37 @@ def __call__(self, envelopecontrol: np.array, audiosample: np.array):
         return amp * signal
 
 
-class Drum:
+class Synth:
+    """
+    An abstract class for a modular synth, ensuring that all modules
+    have the same sample and control rate.
+    """
+
+    def __init__(self, modules: List[SynthModule]):
+        # Check that we are not mixing different control rates or sample rates
+        for m in modules[:1]:
+            assert m.sample_rate == modules[0].sample_rate
+            assert m.control_rate == modules[0].control_rate
+
+
+class Drum(Synth):
     """
     A package of modules that makes one drum hit.
     """
 
     def __init__(
         self,
+        sustain_duration: float,
         pitch_adsr: ADSR = ADSR(),
         amp_adsr: ADSR = ADSR(),
         vco: VCO = VCO(),
         vca: VCA = VCA(),
-        sustain_duration: float = 0,
     ):
+        super().__init__(modules=[pitch_adsr, amp_adsr, vco, vca])
+        # TODO: Should we make this strictly greater than 0.0?
+        assert sustain_duration >= 0
```

Comment:
Why would it ever be negative?

Reply:
0 sustain is fine, for now. We might want to rework the adsr because it's note quite like midi atm, and more setup for drums. can discuss this later.

Reply:
Because it was generated by a neural network? And because these assertions are a useful way for now for documenting way the expected preconditions are for these values?

Reply:
(Because it was generated by a neural network unfamiliar with the API, since we don't have a programmatic way to query the parameter ranges)

Reply:
Ok. Well, either way a 0 is fine. So it can stay >=0

Reply:
Wanna strip that comment?